File: kotlin/src/Terminal.kt
package com.intellipay

import com.fazecast.jSerialComm.*
import com.google.gson.*
import com.google.gson.Gson


class Terminal(_terminalType: String = "v400c-plus", _baudRate: Int = 9600, _byteSize: Int = 8, _stopBit: Int = 1, _xonXoff: Int = 0, _rtsCts: Boolean, _port: String = "/dev/ttyUSB0", _timeout: Int = 0) {
    var terminalType = _terminalType
    var baudRate = _baudRate
    var byteSize = _byteSize
    var stopBit = _stopBit
    var xonXoff = _xonXoff
    var rtsCts = _rtsCts
    var port = _port
    var timeout = _timeout

    fun process( type: String, amount: Double, options: HashMap<String,Any>? ): String {
        val terminal = SerialPort.getCommPort(this.port)
        terminal.setComPortParameters(this.baudRate, this.byteSize, this.stopBit, 0, true)
        if (this.rtsCts)
            terminal.setRTS()
        terminal.setFlowControl(this.xonXoff)
        terminal.setComPortTimeouts(SerialPort.TIMEOUT_READ_BLOCKING, this.timeout, this.timeout)

        val result = HashMap<String, Any>()
        result["status"] = "failed"

        val action = type.toUpperCase()
        if (!listOf("SALE", "REFUND", "VOID").contains(action))
            throw Exception("invalid terminal type [$action]")

        val requestMap: HashMap<String, Any?> = HashMap<String, Any?>()
        requestMap["Amount"] = amount

        requestMap["Operator"] = action;
        if (!options.isNullOrEmpty()) {
            for (key in listOf("Tip", "Invoice", "PaymentItem", "PaymentID")) {
                if (options.containsKey(key)) {
                    requestMap[key] = options[key].toString();
                }
            }
        }
        val gson = Gson()
        val requestString :String = gson.toJson(requestMap)
        val requestByteArray :ByteArray = requestString.toByteArray(Charsets.UTF_8)
        val responseByteArray :ByteArray = ByteArray(255)
        terminal.openPort()
        terminal.writeBytes(requestByteArray, requestByteArray.size.toLong())
        terminal.readBytes(responseByteArray, responseByteArray.size.toLong())
        var responseJson : JsonObject = JsonObject().getAsJsonObject(responseByteArray.toString());
        if (responseJson.has("Response")) {
            responseJson = responseJson.getAsJsonObject("Response")
            if (responseJson.has("PaymentID"))
                result["payment-id"] = responseJson.get("PaymentID").toString()
            if (responseJson.has("ResponseCode") && responseJson.get("ResponseCode").toString() == "00")
                result["status"] = "processed"
            if (responseJson.has("Operation"))
                if (responseJson.get("Operation").toString() == "SALE" || responseJson.get("Operation").toString() == "REFUND")
                    result["card-information"] = hashMapOf(
                            "brand" to responseJson.getAsJsonObject("Card").get("Brand").toString(),
                            "last4" to responseJson.getAsJsonObject("Card").get("Last4").toString(),
                            "expiration" to responseJson.getAsJsonObject("Card").get("ExpDate").toString()
                        )
        }
        terminal.closePort()
        return result.toString()
    }

    fun sale(amount: Double, options: HashMap<String,Any>?): String {
        return process("sale", amount, options )
    }

    fun refund(amount: Double, options: HashMap<String,Any>?): String {
        return process("refund", amount, options )
    }

    fun void (amount: Double, options: HashMap<String,Any>?): String {
        return process("void", amount, options )
    }
}

File: python/IntelliPay/Terminal.py
import serial
import json


class Terminal:

    terminal_type = "v400c-plus"
    terminal = serial.Serial()

    def __init__(self, terminal_type='v400c-plus', baudrate=9600, bytesize=8, stopbit=1, xonxoff=0, rtscts=0, port=4, timeout=0):
        acceptable_terminal_types = ['v200c', 'v400c', 'v400c-plus']

        if terminal_type not in acceptable_terminal_types:
            raise Exception("invalid terminal type [" + terminal_type + "]")

        self.terminal_type = terminal_type
        self.terminal.baudrate = baudrate
        self.terminal.bytesize = bytesize
        self.terminal.stopbit = stopbit
        self.terminal.xonxoff = xonxoff
        self.terminal.rtscts = rtscts
        self.terminal.timeout = timeout
        self.terminal.port = port
        self.terminal.parity = serial.PARITY_NONE

    def prep_message(self, message):
        message = message + str(chr(3))
        lrc = ord(message[0])
        for i in range(1, len(message)):
            lrc ^= ord(message[i])
        message =  str(chr(2)) + message
        prep = message + str(chr(lrc))
        return prep

    def process(self, action, amount, options=None):
        action = action.upper()
        acceptable_action_types = ["SALE", "REFUND", "VOID"]

        if action not in acceptable_action_types:
            raise Exception("invalid terminal action [" + action + "]")

        if action == "VOID" and "PaymentID" not in options:
            raise Exception(action + " requires a PaymentID")

        request_json["Operation"] = action
        request_json = {"Amount": amount}

        if options is not None:
            for field in ["Tip", "Invoice", "Clerk", "PaymentItem", "PaymentID"]:
                if field in options:
                    request_json[field] = options[field]

        request_json = { "Request": request_json }

        # Send request to terminal
        self.terminal.open()
        if (self.terminal.isOpen()):
            print("Connection is open.")

        result = {"status": "failed"}
        response = {}
        if self.terminal.is_open:
            self.terminal.flushInput()
            self.terminal.flushOutput()
            self.terminal.write(bytes(self.prep_message(json.dumps(request_json)), 'ascii'))
            while True:
                data = self.terminal.readline()
                if len(data) > 0:
                    response = json.loads(self.terminal.readline().decode('ascii'))
                    break
            self.terminal.close()
        if "Response" in response:
            response = response["Response"]
            if "PaymentID" in response:
                result["payment-id"] = response["PaymentID"]
            if "ResponseCode" in response and response["ResponseCode"] == "00":
                result["status"] = "processed"  # TODO: talk to matt probably should used "approved" instead.
            if "Operation" in response:
                if response["Operation"] == "SALE" or response["Operation"] == "REFUND":
                    result["card-information"] = dict()
                    result["card-information"]["brand"] = response["Card"]["Brand"]
                    result["card-information"]["last4"] = response["Card"]["Last4"]
                    result["card-information"]["expiration"] = response["Card"]["ExpDate"]
        return result

    def sale(self, amount, options=None):
        return self.process("SALE", amount, options)

    def refund(self, amount, options=None):
        return self.process("REFUND", amount, options)

    def void(self, amount, options=None):
        return self.process("VOID", amount, options)
